### PR TITLE
[WIP] Fix PositionOverlay width issue for the menu in TopBar

### LINKED ISSUE #1673: PositionOverlay incorrectly calculating the left value for right aligned components with scroll bars
## Issue summary

The position overlay is having problems with the calculation of the width of the element. The menu items wrap onto two lines instead of one, and resizing the browser causes all kinds of oddness when the left value is changed to `right: 0;`.

Here is an example of the problem in web:
HbAAMp

## Expected behavior

The menu items do not wrap when there is a scrollbar. In version `4.0.0` we fixed this using the `MutationObserver`.

## Actual behavior

Menu items are wrapping, scroll bar is visible.

## Steps to reproduce the problem

1. Enable view scrollbars, or use firefox
1. Open web and click the user menu in the top right
1. See the menu items wrapping onto two lines

### PR DESCRIPTION
**Pull request by alex-page**
### WHY are these changes introduced?

Fixes #1673 

### WHAT is this pull request doing?

Using mutation observer so that the resize calculation is being run at the correct time. It is also resetting the left value to 0 which is needed for the recalc.

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris-react/blob/master/README.md#development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Versioning%20and%20changelog.md)

This needs to be tested in web and in Polaris with around 6-7 menu items.

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import * as React from 'react';
import {ArrowLeftMinor} from '@shopify/polaris-icons';
import {string} from 'prop-types';
import {
  AppProvider,
  Frame,
  Page,
  Card,
  ActionList,
  TopBar,
  Button,
  Popover,
  PopoverProps,
} from '../src';

interface State {
  userMenuOpen: boolean;
  searchActive: boolean;
  searchText: string;
  popover1Active: boolean;
  popover2Active: boolean;
  popover3Active: boolean;
  popover1PositionBelow: boolean;
  popover2PositionBelow: boolean;
  popover3PositionBelow: boolean;
  showScrollingUserMenu: boolean;
}

export class Playground extends React.Component {
  state: State = {
    userMenuOpen: false,
    searchActive: false,
    searchText: '',
    popover1Active: false,
    popover2Active: false,
    popover3Active: false,
    popover1PositionBelow: true,
    popover2PositionBelow: true,
    popover3PositionBelow: true,
    showScrollingUserMenu: true,
  };

  render() {
    const {
      state,
      handleSearchChange,
      handleSearchResultsDismiss,
      toggleUserMenu,
    } = this;
    const {
      userMenuOpen,
      searchText,
      searchActive,
      showScrollingUserMenu,
    } = state;

    const theme = {
      colors: {
        topBar: {
          background: '#357997',
          backgroundLighter: '#6192a9',
          color: '#FFFFFF',
        },
      },
      logo: {
        width: 124,
        topBarSource:
          'https://cdn.shopify.com/s/files/1/0446/6937/files/jaded-pixel-logo-color.svg?6215648040070010999',
        url: 'http://jadedpixel.com',
        accessibilityLabel: 'Jaded Pixel',
      },
    };

    const actionsWithScroll = [
      {
        items: [{content: 'Back to Shopify', icon: ArrowLeftMinor}],
      },
      {
        items: [{content: 'Community forums'}],
      },
      {
        items: [{content: 'Community forums'}],
      },
      {
        items: [{content: 'Community forums'}],
      },
      {
        items: [{content: 'Community forums'}],
      },
      {
        items: [{content: 'Community forums'}],
      },
      {
        items: [{content: 'Community forums'}],
      },
    ];

    const actions = [
      {
        items: [{content: 'Back to Shopify', icon: ArrowLeftMinor}],
      },
      {
        items: [{content: 'Community forums'}],
      },
      {
        items: [{content: 'Community forums'}],
      },
      {
        items: [{content: 'Community forums'}],
      },
    ];

    const userMenuMarkup = (
      <TopBar.UserMenu
        actions={showScrollingUserMenu ? actionsWithScroll : actions}
        name="Dharma"
        detail="Jaded Pixel"
        initials="D"
        open={userMenuOpen}
        onToggle={toggleUserMenu}
      />
    );

    const searchResultsMarkup = (
      <Card>
        <ActionList
          items={[
            {content: 'Shopify help center'},
            {content: 'Community forums'},
          ]}
        />
      </Card>
    );

    const searchFieldMarkup = (
      <TopBar.SearchField
        onChange={handleSearchChange}
        value={searchText}
        placeholder="Search"
      />
    );

    const topBarMarkup = (
      <TopBar
        showNavigationToggle
        userMenu={userMenuMarkup}
        searchResultsVisible={searchActive}
        searchField={searchFieldMarkup}
        searchResults={searchResultsMarkup}
        onSearchResultsDismiss={handleSearchResultsDismiss}
      />
    );

    const activator1 = (
      <Button onClick={this.togglePopover('popover1', 'Active')}>
        Popover Left
      </Button>
    );

    const activator2 = (
      <Button onClick={this.togglePopover('popover2', 'Active')}>
        Popover Center
      </Button>
    );

    const activator3 = (
      <Button onClick={this.togglePopover('popover3', 'Active')}>
        Popover Right
      </Button>
    );

    return (
      <AppProvider i18n={{}} theme={theme}>
        <Frame topBar={topBarMarkup}>
          <Page
            title="Popover alignment/width bug"
            primaryAction={{
              content: `Render ${
                showScrollingUserMenu ? 'non-scrolling' : 'scrolling'
              } user menu`,
              onAction: this.toggleScrollingUserMenu,
            }}
          >
            <Card>
              <Card.Section
                actions={[
                  {
                    content: 'Toggle position',
                    onAction: this.togglePopover('popover1', 'PositionBelow'),
                  },
                ]}
              >
                <Popover
                  active={this.state.popover1Active}
                  activator={activator1}
                  onClose={this.togglePopover('popover1', 'Active')}
                  preferredAlignment="left"
                  preferredPosition={
                    this.state.popover1PositionBelow ? 'below' : 'above'
                  }
                >
                  <ActionList
                    items={[{content: 'Import'}, {content: 'Export'}]}
                  />
                </Popover>
              </Card.Section>
              <Card.Section
                actions={[
                  {
                    content: 'Toggle position',
                    onAction: this.togglePopover('popover2', 'PositionBelow'),
                  },
                ]}
              >
                <Popover
                  active={this.state.popover2Active}
                  activator={activator2}
                  onClose={this.togglePopover('popover2', 'Active')}
                  preferredAlignment="center"
                  preferredPosition={
                    this.state.popover2PositionBelow ? 'below' : 'above'
                  }
                >
                  <ActionList
                    items={[{content: 'Import'}, {content: 'Export'}]}
                  />
                </Popover>
              </Card.Section>
              <Card.Section
                actions={[
                  {
                    content: 'Toggle position',
                    onAction: this.togglePopover('popover3', 'PositionBelow'),
                  },
                ]}
              >
                <Popover
                  active={this.state.popover3Active}
                  activator={activator3}
                  onClose={this.togglePopover('popover3', 'Active')}
                  preferredAlignment="right"
                  preferredPosition={
                    this.state.popover3PositionBelow ? 'below' : 'above'
                  }
                >
                  <ActionList
                    items={[{content: 'Import'}, {content: 'Export'}]}
                  />
                </Popover>
              </Card.Section>
            </Card>
          </Page>
        </Frame>
      </AppProvider>
    );
  }

  toggleUserMenu = () => {
    this.setState(({userMenuOpen}: State) => ({userMenuOpen: !userMenuOpen}));
  };

  toggleScrollingUserMenu = () => {
    this.setState(({showScrollingUserMenu, userMenuOpen}: State) => ({
      showScrollingUserMenu: !showScrollingUserMenu,
    }));
  };

  togglePopover = (popoverID: string, prop: string) => () => {
    this.setState((state: State) => {
      return {[`${popoverID}${prop}`]: !state[`${popoverID}${prop}`]};
    });
  };

  handleSearchResultsDismiss = () => {
    this.setState(() => {
      return {
        searchActive: false,
        searchText: '',
      };
    });
  };

  handleSearchChange = (value) => {
    this.setState({searchText: value});
    if (value.length > 0) {
      this.setState({searchActive: true});
    } else {
      this.setState({searchActive: false});
    }
  };
}
```

</details>

### 🎩 checklist

* [ ] Tested on [mobile](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md#cross-browser-testing)
* [ ] Tested on [multiple browsers](https://help.shopify.com/en/manual/intro-to-shopify/shopify-admin/supported-browsers)
* [ ] Tested for [accessibility](https://github.com/Shopify/polaris-react/blob/master/documentation/Accessibility%20testing.md)
* [ ] Updated the component's `README.md` with documentation changes
* [ ] [Tophatted documentation](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting%20documentation.md) changes in the style guide

### CODE REVIEW

**Comment by alex-page** [comment]
@ken3650 I have added as a reviewer on this PR as it reverts some of the changes you made in https://github.com/Shopify/polaris-react/pull/1382

@AndrewMusgrave I have added you as a reviewer as it uses your mutation observer code.

**Comment by danrosenthal** [comment]
Can you or @AndrewMusgrave explain about `MutationObserver`?

**Comment by AndrewMusgrave** [comment]
@danrosenthal #1621

If the change is needed for this, I'm tempted to cherry pick the commit into v3

**Comment by danrosenthal** [comment]
> If the change is needed for this, I'm tempted to cherry pick the commit into v3

Seems like it is.

**Comment by alex-page** [comment]
It's not "needed" but it made debugging and performance of the component a lot better.

**Comment by danrosenthal** [comment]
Looks like total code coverage could be achieved with a couple testing tweaks

|:eyes:|
|-|
|Screen Shot 2019-06-18 at 9 57 00 AM|

Testing it additionally by making `fixed` false ought to get us to 90%. Would it also be possible to not have an overlay in one of the tests?

**Comment by AndrewMusgrave** [comment]
> Would it also be possible to not have an overlay in one of the tests?

Nope since the code will never run before `componentDidMount` 😬

**Comment by alex-page** [comment]
This isn't fixed 😭  the issue still exists

**Comment by chloerice** [comment]
Quick update:

I've been trying to narrow down the reason this is an issue, and it seems to only be reproducible with the `UserMenu`. My changes include: 

- Handling `UserMenu` spacing in its parent
  - Currently, its handling its own margin-right, which is causing the activator to be butted up against the far right since its width was including the margin on its wrapper.

- Explicitly setting the `UserMenu` `Popover`, rendered by the `Menu` subcomponent, to have a `preferredPosition` of `right`
  - Currently, no preferred position was specified, so the calculation for center aligned aligned popover was being returned because the conditional inside of `calculateHorizontalPosition` wasn't being satisfied.

- Calculating and setting `right` to horizontally position the overlay when `preferredAlignment="right"` instead calculating and setting `left`

Unfortunately, while I thought I'd fixed this, in spite of these changes the bug is still present in Firefox and Safari :sad

File: src/styles/global.scss [hunk 1]
```diff
@@ -36,6 +36,8 @@ body {
   margin: 0;
   padding: 0;
   background-color: color('sky', 'light');
+  // scrollbar width varies across browsers and operating systems, this value needs to be considered when calculating horizontal alignment of overlays so it must be consistent
+  scrollbar-width: scrollbar-width();
```
Comment (kaelig) [inline src/styles/global.scss]: Length values were [dropped from the spec](https://github.com/w3c/csswg-drafts/commit/3359878276fba119ab68a7bc526c6f205cad289d), the only available values are now `auto | thin | none`.

**Comment by AndrewMusgrave** [comment]
Going to remove myself as a review. Feel free to add me again when this is ready 😄

**Comment by chloerice** [comment]
> Could this PR ship without scrollbar-width?

Setting a consistent `scrollbar-width` isn't helping to fix this issue, and since it's not supported will definitely be removed 👍.

**Comment by alex-page** [comment]
@chloerice This should be in a good state now. Let me know if you want me to help out with TopHat, review etc. 

Lets ship this 👍

**Comment by alex-page** [comment]
Replaces with new PR https://github.com/Shopify/polaris-react/pull/2231

**Pull request closed without merging**